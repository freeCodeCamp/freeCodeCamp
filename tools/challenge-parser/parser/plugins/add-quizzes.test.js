const parseFixture = require('./../__fixtures__/parse-fixture');
const addQuizzes = require('./add-quizzes');

describe('add-quizzes plugin', () => {
  let mockQuizzesAST;
  const plugin = addQuizzes();
  let file = { data: {} };

  beforeAll(async () => {
    mockQuizzesAST = await parseFixture('with-quizzes.md');
  });

  beforeEach(() => {
    file = { data: {} };
  });

  it('returns a function', () => {
    expect(typeof plugin).toEqual('function');
  });

  it('adds a `quizzes` property to `file.data`', () => {
    plugin(mockQuizzesAST, file);

    expect('quizzes' in file.data).toBe(true);
  });

  it('should generate a quizzes array from a quizzes challenge AST', () => {
    expect.assertions(44);
    plugin(mockQuizzesAST, file);

    const quizzes = file.data.quizzes;
    expect(Array.isArray(quizzes)).toBe(true);
    expect(quizzes.length).toBe(2);

    quizzes.forEach(quiz => {
      expect(quiz).toHaveProperty('questions');
      expect(Array.isArray(quiz.questions)).toBe(true);

      const questions = quiz.questions;
      expect(questions.length).toBe(2);

      questions.forEach(question => {
        expect(question).toHaveProperty('text');
        expect(typeof question.text).toBe('string');
        expect(question).toHaveProperty('answer');
        expect(typeof question.answer).toBe('string');
        expect(question).toHaveProperty('distractors');
        expect(Array.isArray(question.distractors)).toBe(true);

        const distractors = question.distractors;

        distractors.forEach(distractor => {
          expect(typeof distractor).toBe('string');
        });
      });
    });
  });

  it('should match the quizzes snapshot', () => {
    plugin(mockQuizzesAST, file);
    expect(file.data).toMatchSnapshot();
  });
});
