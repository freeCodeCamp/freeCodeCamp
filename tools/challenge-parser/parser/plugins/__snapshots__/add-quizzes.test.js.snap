// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add-quizzes plugin should match the quizzes snapshot 1`] = `
{
  "quizzes": [
    {
      "questions": [
        {
          "answer": "<p>Quiz 1, question 1, answer</p>",
          "distractors": [
            "<p>Quiz 1, question 1, distractor 1</p>",
            "<p>Quiz 1, question 1, distractor 2</p>",
            "<p>Quiz 1, question 1, distractor 3</p>",
          ],
          "text": "<p>Quiz 1, question 1</p>",
        },
        {
          "answer": "<p>Quiz 1, question 2, answer</p>",
          "distractors": [
            "<p>Quiz 1, question 2, distractor 1</p>",
            "<p>Quiz 1, question 2, distractor 2</p>",
            "<p>Quiz 1, question 2, distractor 3</p>",
          ],
          "text": "<p>Quiz 1, question 2</p>",
        },
      ],
    },
    {
      "questions": [
        {
          "answer": "<p>Quiz 2, question 1, answer</p>",
          "distractors": [
            "<p>Quiz 2, question 1, distractor 1</p>",
            "<p>Quiz 2, question 1, distractor 2</p>",
            "<p>Quiz 2, question 1, distractor 3</p>",
          ],
          "text": "<p>Quiz 2, question 1</p>",
        },
        {
          "answer": "<p>Quiz 2, question 2, answer</p>",
          "distractors": [
            "<p>Quiz 2, question 2, distractor 1</p>",
            "<p>Quiz 2, question 2, distractor 2</p>",
            "<p>Quiz 2, question 2, distractor 3</p>",
          ],
          "text": "<p>Quiz 2, question 2</p>",
        },
      ],
    },
  ],
}
`;
